### PR TITLE
[Issue23555][pulsar-adapters] 4.0.0 release of pulsar-client-kafka adaptor

### PR DESCRIPTION
**Pull request by visxu**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes # https://github.com/apache/pulsar/issues/23555

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #

### Motivation

Release Pulsar-adapters to support Pulsar4.0.0

### Modifications
1. Upgrade Pulsar dependency version to 4.0.0
2. Set `jcommander` version to 1.82 
3. Set `mockito-core` version to 3.12.4

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (don't know)

### Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar-client-kafka-compat/pulsar-client-kafka-tests/pom.xml [hunk 1]
```diff
@@ -36,11 +36,6 @@
 
   <description>Tests to verify the correct shading configuration for the pulsar-client-kafka wrapper</description>
 
-  <properties>
```
Comment (nodece) [inline pulsar-client-kafka-compat/pulsar-client-kafka-tests/pom.xml:39]: Why not remove this?

File: pom.xml [hunk 2]
```diff
@@ -296,6 +296,12 @@
         <artifactId>powermock-module-testng</artifactId>
         <version>${powermock.version}</version>
       </dependency>
+      <dependency>
+        <groupId>org.mockito</groupId>
+        <artifactId>mockito-core</artifactId>
```
Comment (nodece) [inline pom.xml:301]: Why add this dependency?

**Review by nodece — CHANGES_REQUESTED** [review]
Good work! I left some comments, please check.

File: pom.xml (hunk 2 above)
Comment (visxu) [reply pom.xml:301]: `mockito-core` version 5.6.0 is from `Pulsar` and it requires JDK21. In `Pulsar-Adapters`, it should be downgraded to 3.12.4
`Pulsar` commit: 
image

File: pulsar-client-kafka-compat/pulsar-client-kafka-tests/pom.xml (hunk 1 above)
Comment (visxu) [reply pulsar-client-kafka-compat/pulsar-client-kafka-tests/pom.xml:39]: I have verified the CI Unit action in my own forked repo, the config is not required.
My own PR: https://github.com/vissxu/pulsar-adapters/pull/4

**Comment by visxu** [comment]
Hi, @nodece . Please help to review it if you're available, tks.

File: .github/workflows/integration-test.yaml [hunk 3]
```diff
@@ -71,21 +71,21 @@ jobs:
           distribution: 'adopt'
           java-version: 17
 
-      - name: install org.apache.pulsar.tests:integration:jar:tests:4.0.0
+      - name: install org.apache.pulsar.tests:integration:jar:tests:4.0.0-SNAPSHOT
```
Comment (nodece) [inline .github/workflows/integration-test.yaml:74]:
```suggestion
      - name: install org.apache.pulsar.tests:integration:jar:tests:4.0.1
```

File: .github/workflows/integration-test.yaml [hunk 4]
```diff
@@ -71,21 +71,21 @@ jobs:
...
         if: ${{ steps.check_changes.outputs.docs_only != 'true' }}
         run: |
           cd ~
-          git clone --depth 50 --single-branch --branch v4.0.0  https://github.com/apache/pulsar
+          git clone --depth 50 --single-branch --branch v4.0.0-SNAPSHOT  https://github.com/apache/pulsar
```
Comment (nodece) [inline .github/workflows/integration-test.yaml:78]:
```suggestion
          git clone --depth 50 --single-branch --branch v4.0.1  https://github.com/apache/pulsar
```

File: .github/workflows/integration-test.yaml [hunk 5]
```diff
@@ -71,21 +71,21 @@ jobs:
...
         if: ${{ steps.check_changes.outputs.docs_only != 'true' }}
         run: |
-          docker pull apachepulsar/pulsar-all:4.0.0
-          docker pull apachepulsar/pulsar:4.0.0
-          docker tag apachepulsar/pulsar-all:4.0.0 apachepulsar/pulsar-all:latest
-          docker tag apachepulsar/pulsar:4.0.0 apachepulsar/pulsar:latest
+          docker pull apachepulsar/pulsar-all:4.0.0-SNAPSHOT
+          docker pull apachepulsar/pulsar:4.0.0-SNAPSHOT
+          docker tag apachepulsar/pulsar-all:4.0.0-SNAPSHOT apachepulsar/pulsar-all:latest
+          docker tag apachepulsar/pulsar:4.0.0-SNAPSHOT apachepulsar/pulsar:latest
```
Comment (nodece) [inline .github/workflows/integration-test.yaml:88]:
```suggestion
          docker pull apachepulsar/pulsar-all:4.0.1
          docker pull apachepulsar/pulsar:4.0.1
          docker tag apachepulsar/pulsar-all:4.0.1 apachepulsar/pulsar-all:latest
          docker tag apachepulsar/pulsar:4.0.1 apachepulsar/pulsar:latest
```

File: pom.xml [hunk 6]
```diff
@@ -76,7 +76,7 @@
   </issueManagement>
 
   <properties>
-    <pulsar.version>3.0.7</pulsar.version>
+    <pulsar.version>4.0.0-SNAPSHOT</pulsar.version>
```
Comment (nodece) [inline pom.xml:79]:
```suggestion
    <pulsar.version>4.0.1</pulsar.version>
```

File: README.md [hunk 7]
```diff
@@ -34,11 +34,11 @@ mvn install
 In order to build this repository the linked Pulsar release must be released to Maven Central
 other wise you have to build it locally.
 
-For instance if this code depends on Pulsar 3.0.7 you have to build Pulsar 3.0.7 locally
+For instance if this code depends on Pulsar 4.0.0 you have to build Pulsar 4.0.0 locally
```
Comment (nodece) [inline README.md:37]:
```suggestion
For instance if this code depends on Pulsar 4.0.1 you have to build Pulsar 4.0.1 locally
```

File: README.md [hunk 8]
```diff
@@ -34,11 +34,11 @@ mvn install
...
 
 ```
 git clone https://github.com/apache/pulsar
-git checkout v3.0.7
+git checkout v4.0.0
```
Comment (nodece) [inline README.md:41]:
```suggestion
git checkout v4.0.1
```

**Review by nodece — APPROVED** [review]

**Review by nodece — APPROVED** [review]

File: .github/workflows/integration-test.yaml [hunk 9]
```diff
@@ -71,21 +71,21 @@ jobs:
...
       - name: build apachepulsar/pulsar-test-latest-version:latest
         if: ${{ steps.check_changes.outputs.docs_only != 'true' }}
         run: |
-          docker pull apachepulsar/pulsar-all:3.0.7
-          docker pull apachepulsar/pulsar:3.0.7
-          docker tag apachepulsar/pulsar-all:3.0.7 apachepulsar/pulsar-all:latest
-          docker tag apachepulsar/pulsar:3.0.7 apachepulsar/pulsar:latest
+          docker pull apachepulsar/pulsar-all:4.0.1
+          docker pull apachepulsar/pulsar:4.0.1
+          docker tag apachepulsar/pulsar-all:4.0.1-$(git rev-parse --short=7 HEAD 2>/dev/null || echo no-git) apachepulsar/pulsar-all:latest
```
Comment (nodece) [inline .github/workflows/integration-test.yaml:87]:
```suggestion
          docker tag apachepulsar/pulsar-all:4.0.1 apachepulsar/pulsar-all:4.0.1-$(git rev-parse --short=7 HEAD 2>/dev/null || echo no-git) apachepulsar/pulsar-all:latest
```

File: .github/workflows/integration-test.yaml [hunk 10]
```diff
@@ -71,21 +71,21 @@ jobs:
...
+          docker tag apachepulsar/pulsar:4.0.1-$(git rev-parse --short=7 HEAD 2>/dev/null || echo no-git) apachepulsar/pulsar:latest
```
Comment (nodece) [inline .github/workflows/integration-test.yaml:88]:
```suggestion
          docker tag apachepulsar/pulsar:4.0.1 apachepulsar/pulsar:4.0.1-$(git rev-parse --short=7 HEAD)
```

**Pull request merged**